The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9436724874/job/25991729372) [(plain)](https://github.com/rust-lang/rust/commit/531c1a99f6131ef4cd71a5d08ec4bdaed9a56179/checks/25991729372/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.443   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.460 Collecting boolean-py==4.0
#13 0.467   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.484 Collecting chardet==5.1.0
---
#13 3.617 Building wheels for collected packages: reuse
#13 3.618   Building wheel for reuse (pyproject.toml): started
#13 3.938   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.939   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.939   Stored in directory: /tmp/pip-ephem-wheel-cache-rodj835b/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.942 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.963   Attempting uninstall: setuptools
#13 3.964     Found existing installation: setuptools 59.6.0
#13 3.965     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.965     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.965     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.629 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.630 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.138 Collecting virtualenv
#13 5.191   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.371      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 22.0 MB/s eta 0:00:00
#13 5.424 Collecting platformdirs<5,>=3.9.1
#13 5.432   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.451 Collecting distlib<1,>=0.3.7
#13 5.458   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.505 Collecting filelock<4,>=3.12.2
#13 5.513   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.513   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.599 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.759 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      196544 kB
DirectMap2M:     8192000 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/4f3a276ff1e4e873a739ec34a3ac9e2d365ca2b9/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-4f3a276ff1e4e873a739ec34a3ac9e2d365ca2b9-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 26.26s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs at line 18:
             std: None,
         pointer_width: 32,
         pointer_width: 32,
-        data_layout: "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-i128:128-f64:32:64-f80:32-n8:16:32-S128"
+        data_layout:
+        data_layout:
+            "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-i128:128-f64:32:64-f80:32-n8:16:32-S128"
+                .into(),
         arch: "x86".into(),
         options: base,
     }
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/run-make/manual-crate-name/rmake.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv5te_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/sparc64_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_be_unknown_linux_gnu_ilp32.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_linux_android.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv8r_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_ibm_aix.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64h_apple_darwin.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_nintendo_switch_freestanding.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv4t_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_nto_qnx710.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_teeos.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32gc_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32im_risc0_zkvm_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imc_esp_espidf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_visionos.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unikraft_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7k_apple_watchos.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64el_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/armebv7r_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_uclibceabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64le_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_win7_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_musleabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_pc_nto_qnx700.rs" "/checkout/compiler/rustc_target/src/spec/targets/wasm32_wasi.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_uwp_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_hermit.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_hermit.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_none.rs" "/checkout/compiler/rustc_target/src/spec/targets/hexagon_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv8m_main_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv6k_nintendo_3ds.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_redox.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_kmc_solid_asp3_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv5te_unknown_linux_uclibceabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsisa32r6_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_uwp_windows_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7a_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_none.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imc_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_pc_windows_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/armeb_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/mips64_unknown_linux_gnuabi64.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv8m_base_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/msp430_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64le_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm_unknown_linux_musleabihf.rs" "/checkout/tests/run-make/manual-crate-name/bar.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Sun Jun  9 13:00:08 UTC 2024
  network time: Sun, 09 Jun 2024 13:00:08 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_data_structures/src/flock.rs
+++ b/compiler/rustc_data_structures/src/flock.rs
@@ -9,6 +9,10 @@ cfg_match! {
         mod linux;
         use linux as imp;
     }
+    cfg(target_os = "redox") => {
+        mod linux;
+        use linux as imp;
+    }
     cfg(unix) => {
         mod unix;
         use unix as imp;

--- a/compiler/rustc_target/src/spec/base/redox.rs
+++ b/compiler/rustc_target/src/spec/base/redox.rs
@@ -1,4 +1,4 @@
-use crate::spec::{cvs, RelroLevel, TargetOptions};
+use crate::spec::{cvs, Cc, LinkerFlavor, Lld, RelroLevel, TargetOptions};
 
 pub fn opts() -> TargetOptions {
     TargetOptions {
@@ -12,6 +12,8 @@ pub fn opts() -> TargetOptions {
         has_thread_local: true,
         crt_static_default: true,
         crt_static_respected: true,
+        crt_static_allows_dylibs: true,
+        late_link_args: TargetOptions::link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-lgcc"]),
         ..Default::default()
     }
 }

--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -1647,6 +1647,7 @@ supported_targets! {
     ("x86_64-unknown-l4re-uclibc", x86_64_unknown_l4re_uclibc),
 
     ("aarch64-unknown-redox", aarch64_unknown_redox),
+    ("i686-unknown-redox", i686_unknown_redox),
     ("x86_64-unknown-redox", x86_64_unknown_redox),
 
     ("i386-apple-ios", i386_apple_ios),

--- a/compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs
+++ b/compiler/rustc_target/src/spec/targets/i686_unknown_redox.rs
@@ -1,0 +1,27 @@
+use crate::spec::{base, Cc, LinkerFlavor, Lld, StackProbeType, Target};
+
+pub fn target() -> Target {
+    let mut base = base::redox::opts();
+    base.cpu = "pentiumpro".into();
+    base.plt_by_default = false;
+    base.max_atomic_width = Some(64);
+    base.add_pre_link_args(LinkerFlavor::Gnu(Cc::Yes, Lld::No), &["-m32"]);
+    // don't use probe-stack=inline-asm until rust#83139 and rust#84667 are resolved
+    base.stack_probes = StackProbeType::Call;
+
+    Target {
+        llvm_target: "i686-unknown-redox".into(),
+        metadata: crate::spec::TargetMetadata {
+            description: None,
+            tier: None,
+            host_tools: None,
+            std: None,
+        },
+        pointer_width: 32,
+        data_layout:
+            "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-i128:128-f64:32:64-f80:32-n8:16:32-S128"
+                .into(),
+        arch: "x86".into(),
+        options: base,
+    }
+}

--- a/src/doc/rustc/src/platform-support.md
+++ b/src/doc/rustc/src/platform-support.md
@@ -300,6 +300,7 @@ target | std | host | notes
 [`i686-unknown-hurd-gnu`](platform-support/hurd.md) | ✓ | ✓ | 32-bit GNU/Hurd [^x86_32-floats-return-ABI]
 [`i686-unknown-netbsd`](platform-support/netbsd.md) | ✓ | ✓ | NetBSD/i386 with SSE2 [^x86_32-floats-return-ABI]
 [`i686-unknown-openbsd`](platform-support/openbsd.md) | ✓ | ✓ | 32-bit OpenBSD [^x86_32-floats-return-ABI]
+`i686-unknown-redox` | ? |  | i686 Redox OS
 `i686-uwp-windows-gnu` | ✓ |  | [^x86_32-floats-return-ABI]
 `i686-uwp-windows-msvc` | ✓ |  | [^x86_32-floats-return-ABI]
 [`i686-win7-windows-msvc`](platform-support/win7-windows-msvc.md) | ✓ |   | 32-bit Windows 7 support [^x86_32-floats-return-ABI]

--- a/src/tools/build-manifest/src/main.rs
+++ b/src/tools/build-manifest/src/main.rs
@@ -103,6 +103,7 @@ static TARGETS: &[&str] = &[
     "i686-unknown-freebsd",
     "i686-unknown-linux-gnu",
     "i686-unknown-linux-musl",
+    "i686-unknown-redox",
     "i686-unknown-uefi",
     "loongarch64-unknown-linux-gnu",
     "loongarch64-unknown-none",

--- a/tests/assembly/targets/targets-elf.rs
+++ b/tests/assembly/targets/targets-elf.rs
@@ -228,6 +228,9 @@
 //@ revisions: i686_unknown_openbsd
 //@ [i686_unknown_openbsd] compile-flags: --target i686-unknown-openbsd
 //@ [i686_unknown_openbsd] needs-llvm-components: x86
+//@ revisions: i686_unknown_redox
+//@ [i686_unknown_redox] compile-flags: --target i686-unknown-redox
+//@ [i686_unknown_redox] needs-llvm-components: x86
 //@ revisions: i686_wrs_vxworks
 //@ [i686_wrs_vxworks] compile-flags: --target i686-wrs-vxworks
 //@ [i686_wrs_vxworks] needs-llvm-components: x86

